consolidate error guidance to AIP-193

toumorokoshi opened the pull request:
AIP-193 is the primary AIP around errors. Some
AIPs (primarily those around resource-oriented operations) duplicate some of that guidance.

Consolidating that guidance to help remove redundant text, making it more maintainable.

Also fixing minor formatting incongruences:
  - links not grouped at the bottom.
  - some inconsistent formatting with the changelogs.

toumorokoshi committed:
consolidate error guidance to AIP-193

AIP-193 is the primary AIP around errors. Some
AIPs (primarily those around resource-oriented operations)
duplicate some of that guidance.

Consolidating that guidance to help remove redundant text,
making it more maintanable.

Also fixing minor formatting incongruences:
  - links not grouped at the bottom.
  - some inconsistent formatting with the changelogs.

toumorokoshi commented on aip/general/0193.md:
```diff
@@ -69,13 +69,27 @@ and the method **should** put partial failure information in the metadata
 message. The errors themselves **must** still be represented with a
 [`google.rpc.Status`][Status] object.
 
+### Permission Denied
```
quick note: I think this guidance is actually stronger than the guidance that existed in the original AIPs (since this is more general guidance on all operations, not just resource-oriented ones).

But I think the spirit is the same. Happy to reduce this in scope if we want to discuss the delta in more granular detail.

GregFurmanek commented on aip/general/0131.md:
```diff
@@ -83,24 +83,21 @@ variable in the `google.api.http` annotation on the RPC. This causes the `name`
-whether or not it exists, the service **must** error with `PERMISSION_DENIED`
-(HTTP 403). Permission **must** be checked prior to checking if the resource
-exists.
+See [errors][].
```
You are moving specific section of error handling, perhaps you can link directly to that section and provide related statement in the referencing docs. ex: See "Permission Denied" section in the AIP-193.

toumorokoshi replied:
ack. I wonder if we can just remove the section entirely?

We could just link to permission denied, but the whole of the error guidance is more critical. For example, the standard [payloads include a set of critical requirements on the response returned](https://google.aip.dev/193#guidance).

GregFurmanek replied:
Yes, we could remove the section as long as long as the linking to 193 is properly emphasized.

Also if we are considering the "permission denied" section being the most pertinent for the context
it should also be emphasized even if we are recommending for the reader to review the entire AIP-193.

toumorokoshi replied:
What's your suggestion? I could put this in "further reading"?

GregFurmanek replied:
I don't have an example to point to so for the time being I think keeping a section header would probably be the best unless we come up with an acceptable pattern.

toumorokoshi replied:
ack. Added a specific reference. PTAL!

toumorokoshi committed:
addressing feedback

- adding explicit call-out for permission denied and not found
  errors in CRUD AIPs.

toumorokoshi committed:
Merge branch 'master' into toum/group-error-best-practices

GregFurmanek commented on aip/general/0193.md:
```diff
@@ -69,13 +69,27 @@ and the method **should** put partial failure information in the metadata
+
+If the API requires authentication and authorization, the API **must** first
+check whether the user has the permission to perform the operation, and **must**
+error with `PERMISSION_DENIED` (HTTP 403) if so.
```
I think this could be split into two sentences.

Also phrasing "if so" at the end of the sentence would imply the user has the correct permission in which case the error should not be returned.

shwoodard replied:
@GregFurmanek IIUC, this verbiage was brought over directly from each of the other AIPs. If we want to change it, I would prefer to do that in separate PR.

GregFurmanek replied:
In that case let's use as is and we will update it separately.

shwoodard replied:
Oh wait, scratch that. It was changed.

@toumorokoshi sometimes less is more--leaving some interpretation allowed--is desirable in AIPs. I think this is one of those cases. Can we simply move the repeated copy as is for now? Is there anything specifically required for Uniform Resources that necessitates a change to the existing copy?

toumorokoshi replied:
no, I just like adding context.

I'll revert to the original wording.

GregFurmanek commented on aip/general/0193.md:
```diff
@@ -69,13 +69,27 @@ and the method **should** put partial failure information in the metadata
+check whether the user has the permission to perform the operation, and **must**
+error with `PERMISSION_DENIED` (HTTP 403) if so.
+
+Checking for the permissions and verifying first ensures that the caller cannot
```
I think this can be much simpler.

"Service should return `NOT_FOUND` (HTTP 404) after successful access verification to prevent unauthorized actors from inferring existence of a resource."

shwoodard replied:
Same: IMHO, we should keep the existing text.

toumorokoshi committed:
addressing feedback

- switching back to original wording, but generalizing for
  both parent and resource.

toumorokoshi commented on aip/general/0193.md:
```diff
@@ -69,13 +69,25 @@ and the method **should** put partial failure information in the metadata
+checking if the parent resource exists.
+
+If the user does have proper permission, but the requested parent resource does
+not exist, the service **must** error with `NOT_FOUND` (HTTP 404).
```
@GregFurmanek @shwoodard I think the original text is a little confusing here, since before it was specifically in the context of not found. 

The guidance here would be better off as more general, I think:

- if the user does not have permissions on the resourced, return 403.
- otherwise return the appropriate error as indicated in this AIP.

WDYT?

shwoodard replied:
You have now moved the guidance into AIP-193 so it needs to apply for all Standard methods. In the case of Get and others, there is no parent. Please fix.

toumorokoshi replied:
done, PTAL.

shwoodard commented on aip/general/0134.md:
```diff
@@ -264,28 +264,25 @@ In this situation, an API **may** return back only the fields that were
-exists.
-
-If the user does have proper permission, but the requested resource does not
-exist, the service **must** error with `NOT_FOUND` (HTTP 404) unless
-`allow_missing` is set to `true`.
```
It seems that we lost the bit about `allow_missing` when we consolidated into AIP-193?

toumorokoshi replied:
good catch! re-added to AIP-134 in particular.

toumorokoshi committed:
Merge branch 'master' into toum/group-error-best-practices

shwoodard committed:
Merge branch 'master' into toum/group-error-best-practices
